File: backend/app/schemas/nl2sql.py
# backend/app/schemas/nl2sql.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any


class NL2SQLRequest(BaseModel):
    """NL2SQL 查询请求"""
    question: str = Field(..., description="自然语言问题")
    data_source_id: int = Field(..., description="数据源 ID")
    context: Optional[str] = Field(None, description="上下文信息")
    group_id: Optional[int] = Field(None, description="用户所属集团ID，用于分表选择")


class NL2SQLChartConfig(BaseModel):
    """LLM 推荐图表配置"""
    chart_type: str = Field("bar", description="图表类型: bar, line, pie, scatter")
    x_axis: str = Field("", description="X 轴字段名")
    y_axis: str = Field("", description="Y 轴字段名")
    reason: str = Field("", description="推荐原因")


class GeneratedSQLResult(BaseModel):
    """LLM 生成 SQL 的内部结构化输出"""
    sql: str = Field(..., description="生成的 SELECT SQL 语句")
    confidence: float = Field(0.0, ge=0.0, le=1.0, description="置信度 0-1")
    explanation: str = Field("", description="SQL 生成逻辑说明")
    chart_config: Optional[NL2SQLChartConfig] = Field(None, description="推荐图表配置")


class SQLSuggestion(BaseModel):
    """SQL 建议"""
    sql: str = Field(..., description="生成的 SQL")
    confidence: float = Field(..., description="置信度 0-1")
    explanation: Optional[str] = Field(None, description="解释")
    chart_config: Optional[Dict[str, Any]] = Field(None, description="图表配置：chart_type(图表类型), x_axis(X轴字段), y_axis(Y轴字段)")


class NL2SQLResponse(BaseModel):
    """NL2SQL 查询响应"""
    suggestions: List[SQLSuggestion] = Field(..., description="SQL 建议列表")
    selected_sql: str = Field(..., description="选中的 SQL")
    query_result: Optional[Dict[str, Any]] = Field(None, description="查询结果")
    execution_time_ms: Optional[int] = Field(None, description="执行时间（毫秒）")
    recommended_chart: Optional[Dict[str, Any]] = Field(None, description="推荐的图表配置")

File: frontend/src/api/nl2sql.ts
import request from "@/utils/request"
import type { DataWrapper, PaginatedResponse } from "@/types/api"

export interface FrontendConfig {
  nl2sql_timeout: number
  nl2sql_timeout_ms: number
  llm_adapter: string
  llm_provider: string
  llm_model: string
  llm_api_mode: string
  nl2sql_structured_output_enabled: boolean
  nl2sql_schema_retrieval_enabled: boolean
  nl2sql_schema_retrieval_min_chars: number
  nl2sql_schema_retrieval_max_sections: number
}

export interface NL2SQLRequest {
  question: string
  data_source_id: number
  group_id?: number
  context?: string
}

export interface SQLSuggestion {
  sql: string
  confidence: number
  explanation: string
  chart_config?: {
    chart_type: string
    x_axis: string
    y_axis: string
    reason: string
  }
}

export interface NL2SQLResponse {
  suggestions: SQLSuggestion[]
  selected_sql: string
  query_result?: {
    columns: string[]
    rows: unknown[][]
    total: number
    execution_time_ms: number
  }
  llm_used: boolean
  execution_time_ms: number
}

export interface GroupInfo {
  group_id: number
  group_name: string
}

// ── Cached frontend config ────────────────────────────────

let frontendConfigCache: FrontendConfig | null = null
let configLoaded = false

/**
 * Fetch frontend configuration (NL2SQL timeout, etc.)
 * Cached after first load.
 */
export async function getFrontendConfig(): Promise<FrontendConfig> {
  if (configLoaded && frontendConfigCache) {
    return frontendConfigCache
  }

  try {
    const token = localStorage.getItem('token')
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }
    if (token) {
      headers['Authorization'] = `Bearer ${token}`
    }

    const response = await fetch('/api/config', { method: 'GET', headers })
    if (!response.ok) throw new Error(`HTTP ${response.status}`)

    const data: FrontendConfig = await response.json()
    frontendConfigCache = data
    configLoaded = true
    return data
  } catch (error) {
    console.warn('[NL2SQL] Failed to load frontend config, using defaults:', error)
    const defaults: FrontendConfig = {
      nl2sql_timeout: 300,
      nl2sql_timeout_ms: 360000,
      llm_adapter: 'raw',
      llm_provider: 'openai',
      llm_model: 'gpt-3.5-turbo',
      llm_api_mode: 'chat',
      nl2sql_structured_output_enabled: false,
      nl2sql_schema_retrieval_enabled: true,
      nl2sql_schema_retrieval_min_chars: 12000,
      nl2sql_schema_retrieval_max_sections: 8,
    }
    return defaults
  }
}

/**
 * Get NL2SQL request timeout in milliseconds.
 */
export async function getNL2SQLTimeout(): Promise<number> {
  const config = await getFrontendConfig()
  return config.nl2sql_timeout_ms
}

export function parseQuestion(data: NL2SQLRequest): Promise<NL2SQLResponse> {
  const requestId = `nl2sql_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`
  const startTime = performance.now()

  console.group(`[NL2SQL] 📤 Request | ${requestId}`)
  console.log('├─ data_source_id:', data.data_source_id)
  console.log('├─ question:', data.question)
  console.groupEnd()

  return getNL2SQLTimeout().then((timeout) => {
    return request({
      url: "/nl2sql/parse",
      method: "post",
      data,
      timeout,
    }) as Promise<NL2SQLResponse>
  }).then((response) => {
    const endTime = performance.now()
    console.group(`[NL2SQL] ✅ Success | ${requestId}`)
    console.log('├─ duration:', `${(endTime - startTime).toFixed(2)}ms`)
    console.log('├─ SQL suggestions:', response.suggestions?.length ?? 0)
    console.log('├─ selected SQL:', response.selected_sql)
    console.groupEnd()
    return response
  }).catch((error) => {
    const endTime = performance.now()
    console.group(`[NL2SQL] ❌ Failed | ${requestId}`)
    console.log('├─ duration:', `${(endTime - startTime).toFixed(2)}ms`)
    console.log('├─ error:', error.message || error)
    console.groupEnd()
    throw error
  })
}

/**
 * Get group list for a data source.
 */
export function getGroups(dataSourceId: number): Promise<GroupInfo[]> {
  return request({
    url: "/nl2sql/groups",
    method: "get",
    params: { data_source_id: dataSourceId }
  }).then((response: unknown) => {
    const res = response as DataWrapper<GroupInfo[]>
    return res.data || []
  })
}
